test/controllers/incomeSources/cease/IncomeSourceEndDateControllerSpec.scala

```diff
@@ -126,669 +127,426 @@ class IncomeSourceEndDateControllerSpec extends TestSupport with MockAuthenticat
-              "income-source-end-date.year" -> "2022"))
+        val result: Future[Result] = (isAgent, isChange) match {
+          case (true, true) =>
+            TestIncomeSourceEndDateController.submitChangeAgent(id, incomeSourceType)(fakeRequestConfirmedClient().withDateInFormEncoding)
+          case (true, false) =>
+            TestIncomeSourceEndDateController.submitAgent(id, incomeSourceType)(fakeRequestConfirmedClient().withDateInFormEncoding)
+          case (false, true) =>
+            TestIncomeSourceEndDateController.submitChange(id, incomeSourceType)(fakeRequestWithActiveSession.withDateInFormEncoding)
+          case (false, false) =>
+            TestIncomeSourceEndDateController.submit(id, incomeSourceType)(fakeRequestWithActiveSession.withDateInFormEncoding)
         }
 
-        status(result) shouldBe Status.BAD_REQUEST
-      }
-    }
-    "return 500 INTERNAL SERVER ERROR to internal server page" when {
-      "income source ID is missing and income source type is Self Employment " in {
-        setupMockAgentAuthRetrievalSuccess(testAgentAuthRetrievalSuccess)
-        enable(IncomeSources)
-        mockBusinessIncomeSource()
-        val incomeSourceType = SelfEmployment
-        lazy val result: Future[Result] = {
-          TestIncomeSourceEndDateController.submitChangeAgent(None, incomeSourceType)(fakeRequestConfirmedClient()
-            .withFormUrlEncodedBody("income-source-end-date.day" -> "27", "income-source-end-date.month" -> "8",
-              "income-source-end-date.year" -> "2022"))
-        }
         status(result) shouldBe INTERNAL_SERVER_ERROR
       }
 
-      "income source ID is missing" in {
-        setupMockAgentAuthRetrievalSuccess(testAgentAuthRetrievalSuccess)
-        enable(IncomeSources)
-        mockBusinessIncomeSource()
-        lazy val result: Future[Result] = {
-          TestIncomeSourceEndDateController.submitChangeAgent(None, incomeSourceType = SelfEmployment)(fakeRequestConfirmedClient()
-            .withFormUrlEncodedBody("income-source-end-date.day" -> "27", "income-source-end-date.month" -> "8",
-              "income-source-end-date.year" -> "2022"))
+      "income source ID is missing and income source type is Self Employment " when {
+        "called .submit" when {
+          "user is an Individual" in {
+            testMissingIncomeSourceID(isAgent = false, isChange = false)
+          }
+          "user is an Agent" in {
+            testMissingIncomeSourceID(isAgent = true, isChange = false)
+          }
+        }
+        "called .submitChange" when {
+          "user is an Individual" in {
+            testMissingIncomeSourceID(isAgent = false, isChange = true)
+          }
+          "user is an Agent" in {
+            testMissingIncomeSourceID(isAgent = true, isChange = true)
+          }
         }
-        status(result) shouldBe INTERNAL_SERVER_ERROR
       }
     }
   }
+
```

Comment:
Could fix: add checks for returning 500 when Mongo returns an Exception
(code coverage is already 94.56% though)